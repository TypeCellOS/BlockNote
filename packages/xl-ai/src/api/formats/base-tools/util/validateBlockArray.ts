import { DeepPartial } from "ai";
import { Result } from "../../../../streamTool/streamTool.js";

export function validateBlockArray<U>(
  inputArray: DeepPartial<Array<U>>,
  validateItem: (item: DeepPartial<U>) => Result<U>,
): Result<U[]> {
  if (!inputArray || !Array.isArray(inputArray) || inputArray.length === 0) {
    return {
      ok: false,
      error: "blocks is required",
    };
  }

  const validatedBlocks: U[] = [];

  for (const item of inputArray) {
    const validationResult = validateItem(item as DeepPartial<U>);
    if (!validationResult.ok) {
      return {
        ok: false,
        error: `Invalid block: ${validationResult.error}`,
      };
    }
    validatedBlocks.push(validationResult.value);
  }

  return {
    ok: true,
    value: validatedBlocks,
  };
}
