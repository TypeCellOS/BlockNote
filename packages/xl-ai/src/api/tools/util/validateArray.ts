import { DeepPartial } from "ai";
import { InvalidOrOk } from "../../streamTool/streamTool.js";

export function validateArray<U>(
  inputArray: DeepPartial<Array<U>>,
  validateItem: (item: DeepPartial<U>) => InvalidOrOk<U>
): InvalidOrOk<U[]> {
  if (!inputArray || !Array.isArray(inputArray) || inputArray.length === 0) {
    return {
      result: "invalid",
      reason: "blocks is required",
    };
  }

  const validatedBlocks: U[] = [];

  for (const item of inputArray) {
    const validationResult = validateItem(item as DeepPartial<U>);
    if (validationResult.result === "invalid") {
      return {
        result: "invalid",
        reason: `Invalid block: ${validationResult.reason}`,
      };
    }
    validatedBlocks.push(validationResult.value);
  }

  return {
    result: "ok",
    value: validatedBlocks,
  };
}
