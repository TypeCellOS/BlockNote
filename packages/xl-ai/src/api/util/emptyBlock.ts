import { Block } from "@blocknote/core";

// TODO: move?
export function isEmptyParagraph(block: Block<any, any, any>) {
    return (
      block.type === "paragraph" &&
      Array.isArray(block.content) &&
      block.content.length === 0
    );
  }