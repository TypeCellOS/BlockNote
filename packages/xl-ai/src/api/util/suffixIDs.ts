import type {
  Block,
  BlockSchema,
  InlineContentSchema,
  StyleSchema,
} from "@blocknote/core";

export function suffixIDs<
  T extends Block<BlockSchema, InlineContentSchema, StyleSchema>
>(blocks: T[]): T[] {
  return blocks.map((block) => ({
    ...block,
    id: `${block.id}$`,
    children: suffixIDs(block.children),
  }));
}
