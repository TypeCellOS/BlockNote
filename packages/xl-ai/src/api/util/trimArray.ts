export function trimArray<T>(
  arr: T[],
  matchFn: (element: T) => boolean,
  trimStart = true,
  trimEnd = true
): T[] {
  let start = 0;
  let end = arr.length;

  // Find the first non-matching element from the start
  if (trimStart) {
    while (start < end && matchFn(arr[start])) {
      start++;
    }
  }

  // Find the first non-matching element from the end
  if (trimEnd) {
    while (end > start && matchFn(arr[end - 1])) {
      end--;
    }
  }

  // Slice the array to include only the untrimmed portion
  return arr.slice(start, end);
}
