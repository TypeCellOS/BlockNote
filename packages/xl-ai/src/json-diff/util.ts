export const extendedTypeOf = function (obj: any) {
  const result = typeof obj;
  if (obj == null) {
    return "null";
  } else if (result === "object" && obj.constructor === Array) {
    return "array";
  } else if (result === "object" && obj instanceof Date) {
    return "date";
  } else {
    return result;
  }
};

export const roundObj = function (data: any, precision: number): any {
  const type = typeof data;
  if (Array.isArray(data)) {
    return data.map((x: any) => roundObj(x, precision));
  } else if (type === "object") {
    for (const key in data) {
      data[key] = roundObj(data[key], precision);
    }
    return data;
  } else if (
    type === "number" &&
    Number.isFinite(data) &&
    !Number.isInteger(data)
  ) {
    return +data.toFixed(precision);
  } else {
    return data;
  }
};
