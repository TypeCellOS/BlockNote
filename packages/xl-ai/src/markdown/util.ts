import remarkStringify from "remark-stringify";
import { unified } from "unified";

export function markdownNodeToString(node: any) {
  const md = unified().use(remarkStringify).stringify(node);
  return md;
}
