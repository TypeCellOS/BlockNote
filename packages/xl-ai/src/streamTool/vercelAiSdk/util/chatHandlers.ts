import { getErrorMessage } from "@ai-sdk/provider-utils";
import { Chat } from "@ai-sdk/react";
import { DeepPartial, isToolUIPart, UIMessage } from "ai";
import { StreamTool, StreamToolCall } from "../../streamTool.js";
import {
  ChunkExecutionError,
  StreamToolExecutor,
} from "../../StreamToolExecutor.js";
import { objectStreamToOperationsResult } from "./UIMessageStreamToOperationsResult.js";

/**
 * Listens to messages received in the `chat` object and processes tool calls
 * by streaming them to an executor
 *
 * It also listens to the status and error events of the chat object and handles them
 * appropriately.
 *
 * It also waits for all tool calls to be completed and then adds the results to the chat object.
 *
 * NOTE: listening to the `chat` object + error handling is a bit cumbersome. It might have been
 * cleaner to directly listen to the UIMessageStream. However, for that we'd probably
 * need to wrap the transport or chat object in AIExtension
 *
 * The error handling is currently quite convoluted. To properly test this,
 * you can:
 * a) make sure a tool call fails
 * b) make sure the entire request fails (network error)
 *
 */
export async function setupToolCallStreaming(
  streamTools: StreamTool<any>[],
  chat: Chat<UIMessage>,
  onStart?: () => void,
) {
  /*
  We use a single executor even for multiple tool calls.
  This is because a tool call operation (like Add), might behave differently
  if a block has been added earlier (i.e.: executing tools can keep state, 
  and this state is shared across parallel tool calls).
  */
  const executor = new StreamToolExecutor(streamTools);

  const appendableStream = createAppendableStream<any>();

  appendableStream.output.pipeTo(executor.writable);

  const toolCallStreams = new Map<string, ToolCallStreamData>();

  let first = true;

  // Possible improvement: instead of pushing tool call updates directly,
  // we could make this a readablestream where we return the latest state
  // of the tool call input. This way we don't process all intermediate
  // streaming steps in case downstream consumer (like the StreamToolExecutor)
  // are slower than the producer
  const unsub = chat["~registerMessagesCallback"](() => {
    processToolCallParts(chat, (data) => {
      if (!toolCallStreams.has(data.toolCallId)) {
        const toolCallStreamData = createToolCallStream(
          streamTools,
          data.toolName,
          data.toolCallId,
        );
        appendableStream.append(toolCallStreamData.operationsStream);
        toolCallStreams.set(data.toolCallId, toolCallStreamData);
        if (first) {
          first = false;
          onStart?.();
        }
      }
      return toolCallStreams.get(data.toolCallId)!;
    });
  });

  const statusHandler = new Promise<void>((resolve) => {
    const unsub2 = chat["~registerStatusCallback"](() => {
      if (chat.status === "ready" || chat.status === "error") {
        unsub();
        unsub2();
        if (chat.status !== "error") {
          // don't unsubscribe the error listener if chat.status === "error"
          // we need to wait for the error event, because only in the error event we can read chat.error
          // (in this status listener, it's still undefined)
          unsub3();
        }
        resolve();
      }
    });

    const unsub3 = chat["~registerErrorCallback"](() => {
      if (chat.error) {
        unsub3();
        for (const data of toolCallStreams.values()) {
          if (!data.complete) {
            // this can happen in case of a network error for example
            data.writer.abort(chat.error);
          }
        }
        // reject(chat.error);
        // we intentionally commented out the above line to not reject here
        // instead, we abort (raise an error) in the unfinished tool calls
      }
    });
  });

  // wait until all messages have been received
  // (possible improvement(?): we can abort the request if any of the tool calls fail
  //  instead of waiting for the entire llm response)
  await statusHandler;

  // we're not going to append any more streams from tool calls, because we've seen all tool calls
  await appendableStream.finalize();
  // let all stream executors finish, this can take longer due to artificial delays
  // (e.g. to simulate human typing behaviour)
  const result = (await Promise.allSettled([executor.finish()]))[0];

  let error: ChunkExecutionError | undefined;
  if (result.status === "rejected") {
    if (result.reason instanceof ChunkExecutionError) {
      error = result.reason;
    } else {
      if (!chat.error) {
        throw new Error(
          "Unexpected: no ChunkExecutionError but also no chat.error (network error?)",
        );
      }
    }
  }

  let errorSeen = false;
  // process results
  const toolCalls = Array.from(
    toolCallStreams.values().filter((data) => data.complete),
  );
  toolCalls.forEach((toolCall, index) => {
    const isErrorTool =
      toolCall.toolCallId === error?.chunk.metadata.toolCallId;

    if (isErrorTool) {
      errorSeen = true;
    }

    chat.addToolResult({
      tool: toolCalls[index].toolName,
      toolCallId: toolCalls[index].toolCallId,
      output:
        errorSeen === false
          ? { status: "ok" }
          : isErrorTool
            ? { status: "error", error: getErrorMessage(error) }
            : { status: "not-executed-previous-tool-errored" },
    });
  });

  if (error) {
    throw error;
  }

  if (chat.error) {
    // response failed
    throw chat.error;
  }
}

function createAppendableStream<T>() {
  let controller: ReadableStreamDefaultController<T>;
  let ready = Promise.resolve();
  let canceled = false;
  const output = new ReadableStream({
    start(c) {
      controller = c;
    },
    cancel(reason) {
      canceled = true;
      controller.error(reason);
    },
  });

  async function append(readable: ReadableStream<T>) {
    if (canceled) {
      throw new Error("Appendable stream canceled, can't append");
    }
    const reader = readable.getReader();

    // Chain appends in sequence
    ready = ready.then(async () => {
      // eslint-disable-next-line no-constant-condition
      while (true) {
        try {
          const { done, value } = await reader.read();
          if (done || canceled) {
            break;
          }
          controller.enqueue(value);
        } catch (e) {
          canceled = true;
          controller.error(e);
          break;
        }
      }
    });

    return ready;
  }

  async function finalize() {
    await ready; // wait for last appended stream to finish

    if (!canceled) {
      controller.close(); // only close once no more streams will come
    }
  }

  return { output, append, finalize };
}

// Types for tool call streaming
type ToolCallStreamData = {
  // stream: TransformStream<DeepPartial<{ operations: StreamToolCall<any>[] }>>;
  writer: WritableStreamDefaultWriter<
    DeepPartial<{ operations: StreamToolCall<any>[] }>
  >;
  complete: boolean;
  toolName: string;
  toolCallId: string;
  operationsStream: ReadableStream<any>;
  // executor: StreamToolExecutor<any>;
};

/**
 * Process tool call parts and manage individual streams per toolCallId
 */
function processToolCallParts(
  chat: Chat<UIMessage>,
  getToolCallStreamData: (data: {
    toolName: string;
    toolCallId: string;
  }) => ToolCallStreamData,
) {
  for (const part of chat.lastMessage?.parts ?? []) {
    if (!isToolUIPart(part)) {
      continue;
    }

    const toolName = part.type.replace("tool-", "");

    if (toolName !== "applyDocumentOperations") {
      // we only process the combined operations tool call
      // in a future improvement we can add more generic support for different tool streaming
      continue;
    }

    const toolCallId = part.toolCallId;

    const toolCallData = getToolCallStreamData({
      toolName: part.type.replace("tool-", ""),
      toolCallId,
    });

    processToolCallPart(part, toolCallData);
  }
}

/**
 * Create a complete stream pipeline for a single toolCallId
 */
function createToolCallStream(
  streamTools: StreamTool<any>[],
  toolName: string,
  toolCallId: string,
): ToolCallStreamData {
  const stream = new TransformStream<
    DeepPartial<{ operations: StreamToolCall<any>[] }>
  >();
  const operationsStream = objectStreamToOperationsResult(
    stream.readable,
    streamTools,
    { toolCallId },
  );

  const writer = stream.writable.getWriter();

  return {
    // stream,
    writer,
    complete: false,
    // executor,
    operationsStream,
    toolName,
    toolCallId,
  };
}

/**
 * Process a single tool call part (streaming or available)
 */
function processToolCallPart(part: any, toolCallData: ToolCallStreamData) {
  if (part.state === "input-streaming") {
    const input = part.input;
    if (input !== undefined) {
      toolCallData.writer.write(input as any);
    }
  } else if (part.state === "input-available") {
    const input = part.input;
    if (input === undefined) {
      throw new Error("input is undefined");
    }
    if (!toolCallData.complete) {
      toolCallData.complete = true;
      toolCallData.writer.write(input as any);
      toolCallData.writer.close();
    }
  }
}
