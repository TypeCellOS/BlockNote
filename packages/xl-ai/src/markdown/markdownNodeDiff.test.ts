import { describe, expect, it } from "vitest";
import {
  MarkdownNodeDiffResult,
  markdownNodeDiff,
} from "./markdownNodeDiff.js";
import { markdownNodeToString } from "./util.js";

async function operationsToReadableString(
  operations: MarkdownNodeDiffResult[]
) {
  return operations
    .map((op) => {
      const md = markdownNodeToString(
        op.type === "remove" ? op.oldBlock : op.newBlock
      );
      return `${op.type.toUpperCase()}: ${md}`;
    })
    .join("\n");
}

describe("markdown AI util", () => {
  it("collapse lines", async () => {
    const md1 = `# title

hello`;

    const md2 = `# titleworldhello`;

    const operations = await markdownNodeDiff(md1, md2);
    const diff = await operationsToReadableString(operations);
    expect(diff).toMatchSnapshot();
  });

  it("expand lines", async () => {
    const md1 = `# titleworldhello`;

    const md2 = `# title

hello`;

    const operations = await markdownNodeDiff(md1, md2);
    const diff = await operationsToReadableString(operations);
    expect(diff).toMatchSnapshot();
  });

  it("add sentence at end", async () => {
    const md1 = `# title

hello`;

    const md2 = `# title

hello

world`;

    const operations = await markdownNodeDiff(md1, md2);
    const diff = await operationsToReadableString(operations);
    expect(diff).toMatchSnapshot();
  });

  it("remove sentence at end", async () => {
    const md1 = `# title

hello

world`;

    const md2 = `# title

hello`;

    const operations = await markdownNodeDiff(md1, md2);
    const diff = await operationsToReadableString(operations);
    expect(diff).toMatchSnapshot();
  });

  it("add sentence at start", async () => {
    const md1 = `# title

hello

world`;

    const md2 = `prefix
# title

hello

world`;

    const operations = await markdownNodeDiff(md1, md2);
    const diff = await operationsToReadableString(operations);
    expect(diff).toMatchSnapshot();
  });

  it("remove sentence at start", async () => {
    const md1 = `# title
    
hello

world`;

    const md2 = `hello

world`;

    const operations = await markdownNodeDiff(md1, md2);
    const diff = await operationsToReadableString(operations);
    expect(diff).toMatchSnapshot();
  });

  it("add sentence at middle", async () => {
    const md1 = `# title

hello

world`;

    const md2 = `# title

hello

beautiful

world`;

    const operations = await markdownNodeDiff(md1, md2);
    const diff = await operationsToReadableString(operations);
    expect(diff).toMatchSnapshot();
  });

  it("remove sentence at middle", async () => {
    const md1 = `# title

hello

world`;

    const md2 = `# title

world`;

    const operations = await markdownNodeDiff(md1, md2);
    const diff = await operationsToReadableString(operations);
    expect(diff).toMatchSnapshot();
  });

  it("change: type at start", async () => {
    const md1 = `# title

hello

world`;

    const md2 = `title

hello

wold`;

    const operations = await markdownNodeDiff(md1, md2);
    const diff = await operationsToReadableString(operations);
    expect(diff).toMatchSnapshot();
  });

  it("change: type at end", async () => {
    const md1 = `# title

hello

world`;

    const md2 = `# title

hello

# world`;

    const operations = await markdownNodeDiff(md1, md2);
    const diff = await operationsToReadableString(operations);
    expect(diff).toMatchSnapshot();
  });

  it("change: text at end", async () => {
    const md1 = `# title

hello

world`;

    const md2 = `# title

hello

worlds`;

    const operations = await markdownNodeDiff(md1, md2);
    const diff = await operationsToReadableString(operations);
    expect(diff).toMatchSnapshot();
  });

  it("change: text at start", async () => {
    const md1 = `# new title

hello

world`;

    const md2 = `# title

hello

world`;

    const operations = await markdownNodeDiff(md1, md2);
    const diff = await operationsToReadableString(operations);
    expect(diff).toMatchSnapshot();
  });

  it("change: text at middle", async () => {
    const md1 = `# title

hello

world`;

    const md2 = `# title

hello there,

world`;

    const operations = await markdownNodeDiff(md1, md2);
    const diff = await operationsToReadableString(operations);
    expect(diff).toMatchSnapshot();
  });

  it("mixed: change and insert", async () => {
    const md1 = `# title

hello

world`;

    const md2 = `# title

hello there

beautiful

wold`;

    const operations = await markdownNodeDiff(md1, md2);
    const diff = await operationsToReadableString(operations);
    expect(diff).toMatchSnapshot();
  });

  it("lists: remove list item", async () => {
    const md1 = `# title

- hello
- world`;

    const md2 = `# title

- hello`;

    const operations = await markdownNodeDiff(md1, md2);
    const diff = await operationsToReadableString(operations);
    expect(diff).toMatchSnapshot();
  });

  it("lists: add list item", async () => {
    const md1 = `# title

- hello`;

    const md2 = `# title

- hello
- world`;

    const operations = await markdownNodeDiff(md1, md2);
    const diff = await operationsToReadableString(operations);
    expect(diff).toMatchSnapshot();
  });

  it("lists: update list item", async () => {
    const md1 = `# title

- hello
- world`;

    const md2 = `# title

- hello there
- world`;

    const operations = await markdownNodeDiff(md1, md2);
    const diff = await operationsToReadableString(operations);
    expect(diff).toMatchSnapshot();
  });

  it("lists: add list", async () => {
    const md1 = `# title`;

    const md2 = `# title

- hello there
- world`;

    const operations = await markdownNodeDiff(md1, md2);
    const diff = await operationsToReadableString(operations);
    expect(diff).toMatchSnapshot();
  });

  it("lists: remove list", async () => {
    const md1 = `# title

- hello there
- world`;

    const md2 = `# title`;

    const operations = await markdownNodeDiff(md1, md2);
    const diff = await operationsToReadableString(operations);
    expect(diff).toMatchSnapshot();
  });

  it("3 lines -> large text", async () => {
    const md1 = `# title

hello

world`;

    const md2 =
      "# Document Title\n\nTest\n\nThis is the first sentence of the document. \n\n## Introduction\n\nThe introduction provides an overview of the document's purpose and structure. \n\n## Main Content\n\nThe main content section contains the bulk of the information. \n\n### Subsection 1\n\nDetails about the first subsection are provided here. \n\n### Subsection 2\n\nDetails about the second subsection are provided here. \n\n## Conclusion\n\nThe conclusion summarizes the key points discussed in the document. \n\n## References\n\nA list of references and resources used in the document.";

    const operations = await markdownNodeDiff(md1, md2);
    const diff = await operationsToReadableString(operations);
    expect(diff).toMatchSnapshot();
  });

  it("1 line -> large text", async () => {
    const md1 = `hello`;

    const md2 =
      "# Document Title\n\nTest\n\nThis is the first sentence of the document. \n\n## Introduction\n\nThe introduction provides an overview of the document's purpose and structure. \n\n## Main Content\n\nThe main content section contains the bulk of the information. \n\n### Subsection 1\n\nDetails about the first subsection are provided here. \n\n### Subsection 2\n\nDetails about the second subsection are provided here. \n\n## Conclusion\n\nThe conclusion summarizes the key points discussed in the document. \n\n## References\n\nA list of references and resources used in the document.";

    const operations = await markdownNodeDiff(md1, md2);
    const diff = await operationsToReadableString(operations);
    expect(diff).toMatchSnapshot();
  });

  it("large text -> 3 lines", async () => {
    const md1 =
      "# Document Title\n\nTest\n\nThis is the first sentence of the document. \n\n## Introduction\n\nThe introduction provides an overview of the document's purpose and structure. \n\n## Main Content\n\nThe main content section contains the bulk of the information. \n\n### Subsection 1\n\nDetails about the first subsection are provided here. \n\n### Subsection 2\n\nDetails about the second subsection are provided here. \n\n## Conclusion\n\nThe conclusion summarizes the key points discussed in the document. \n\n## References\n\nA list of references and resources used in the document.";

    const md2 = `# title

hello

world`;

    const operations = await markdownNodeDiff(md1, md2);
    const diff = await operationsToReadableString(operations);
    expect(diff).toMatchSnapshot();
  });

  it("large text -> 1 line", async () => {
    const md1 =
      "# Document Title\n\nTest\n\nThis is the first sentence of the document. \n\n## Introduction\n\nThe introduction provides an overview of the document's purpose and structure. \n\n## Main Content\n\nThe main content section contains the bulk of the information. \n\n### Subsection 1\n\nDetails about the first subsection are provided here. \n\n### Subsection 2\n\nDetails about the second subsection are provided here. \n\n## Conclusion\n\nThe conclusion summarizes the key points discussed in the document. \n\n## References\n\nA list of references and resources used in the document.";

    const md2 = `hello`;
    const operations = await markdownNodeDiff(md1, md2);
    const diff = await operationsToReadableString(operations);
    expect(diff).toMatchSnapshot();
  });

  it("3 lines -> 1 line", async () => {
    const md1 = "# Document Title\n\nTest\n\nThis";

    const md2 = `mestis`;

    const operations = await markdownNodeDiff(md1, md2);
    const diff = await operationsToReadableString(operations);
    expect(diff).toMatchSnapshot();
  });
});
