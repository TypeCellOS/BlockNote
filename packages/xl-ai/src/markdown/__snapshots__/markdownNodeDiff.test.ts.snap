// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`markdown AI util > 1 line -> large text 1`] = `
"ADD: # Document Title

ADD: Test

CHANGED: This is the first sentence of the document.

ADD: ## Introduction

ADD: The introduction provides an overview of the document's purpose and structure.

ADD: ## Main Content

ADD: The main content section contains the bulk of the information.

ADD: ### Subsection 1

ADD: Details about the first subsection are provided here.

ADD: ### Subsection 2

ADD: Details about the second subsection are provided here.

ADD: ## Conclusion

ADD: The conclusion summarizes the key points discussed in the document.

ADD: ## References

ADD: A list of references and resources used in the document.
"
`;

exports[`markdown AI util > 3 lines -> 1 line 1`] = `
"CHANGED: mestis

REMOVE: Test

REMOVE: This
"
`;

exports[`markdown AI util > 3 lines -> large text 1`] = `
"CHANGED: # Document Title

ADD: Test

CHANGED: This is the first sentence of the document.

ADD: ## Introduction

ADD: The introduction provides an overview of the document's purpose and structure.

ADD: ## Main Content

ADD: The main content section contains the bulk of the information.

ADD: ### Subsection 1

ADD: Details about the first subsection are provided here.

CHANGED: ### Subsection 2

ADD: Details about the second subsection are provided here.

ADD: ## Conclusion

ADD: The conclusion summarizes the key points discussed in the document.

ADD: ## References

ADD: A list of references and resources used in the document.
"
`;

exports[`markdown AI util > add sentence at end 1`] = `
"UNCHANGED: # title

UNCHANGED: hello

ADD: world
"
`;

exports[`markdown AI util > add sentence at middle 1`] = `
"UNCHANGED: # title

UNCHANGED: hello

ADD: beautiful

UNCHANGED: world
"
`;

exports[`markdown AI util > add sentence at start 1`] = `
"ADD: prefix

UNCHANGED: # title

UNCHANGED: hello

UNCHANGED: world
"
`;

exports[`markdown AI util > change: text at end 1`] = `
"UNCHANGED: # title

UNCHANGED: hello

CHANGED: worlds
"
`;

exports[`markdown AI util > change: text at middle 1`] = `
"UNCHANGED: # title

CHANGED: hello there,

UNCHANGED: world
"
`;

exports[`markdown AI util > change: text at start 1`] = `
"CHANGED: # title

UNCHANGED: hello

UNCHANGED: world
"
`;

exports[`markdown AI util > change: type at end 1`] = `
"UNCHANGED: # title

UNCHANGED: hello

CHANGED: # world
"
`;

exports[`markdown AI util > change: type at start 1`] = `
"CHANGED: title

UNCHANGED: hello

CHANGED: wold
"
`;

exports[`markdown AI util > collapse lines 1`] = `
"CHANGED: # titleworldhello

REMOVE: hello
"
`;

exports[`markdown AI util > expand lines 1`] = `
"CHANGED: # title

ADD: hello
"
`;

exports[`markdown AI util > large text -> 1 line 1`] = `
"REMOVE: # Document Title

REMOVE: Test

CHANGED: hello

REMOVE: ## Introduction

REMOVE: The introduction provides an overview of the document's purpose and structure.

REMOVE: ## Main Content

REMOVE: The main content section contains the bulk of the information.

REMOVE: ### Subsection 1

REMOVE: Details about the first subsection are provided here.

REMOVE: ### Subsection 2

REMOVE: Details about the second subsection are provided here.

REMOVE: ## Conclusion

REMOVE: The conclusion summarizes the key points discussed in the document.

REMOVE: ## References

REMOVE: A list of references and resources used in the document.
"
`;

exports[`markdown AI util > large text -> 3 lines 1`] = `
"CHANGED: # title

REMOVE: Test

CHANGED: hello

REMOVE: ## Introduction

REMOVE: The introduction provides an overview of the document's purpose and structure.

REMOVE: ## Main Content

REMOVE: The main content section contains the bulk of the information.

REMOVE: ### Subsection 1

REMOVE: Details about the first subsection are provided here.

CHANGED: world

REMOVE: Details about the second subsection are provided here.

REMOVE: ## Conclusion

REMOVE: The conclusion summarizes the key points discussed in the document.

REMOVE: ## References

REMOVE: A list of references and resources used in the document.
"
`;

exports[`markdown AI util > lists: add list 1`] = `
"UNCHANGED: # title

ADD: *   hello there

ADD: *   world
"
`;

exports[`markdown AI util > lists: add list item 1`] = `
"UNCHANGED: # title

UNCHANGED: *   hello

ADD: *   world
"
`;

exports[`markdown AI util > lists: remove list 1`] = `
"UNCHANGED: # title

REMOVE: *   hello there

REMOVE: *   world
"
`;

exports[`markdown AI util > lists: remove list item 1`] = `
"UNCHANGED: # title

UNCHANGED: *   hello

REMOVE: *   world
"
`;

exports[`markdown AI util > lists: update list item 1`] = `
"UNCHANGED: # title

CHANGED: *   hello there

UNCHANGED: *   world
"
`;

exports[`markdown AI util > mixed: change and insert 1`] = `
"UNCHANGED: # title

CHANGED: hello there

ADD: beautiful

CHANGED: wold
"
`;

exports[`markdown AI util > remove sentence at end 1`] = `
"UNCHANGED: # title

UNCHANGED: hello

REMOVE: world
"
`;

exports[`markdown AI util > remove sentence at middle 1`] = `
"UNCHANGED: # title

REMOVE: hello

UNCHANGED: world
"
`;

exports[`markdown AI util > remove sentence at start 1`] = `
"REMOVE: # title

UNCHANGED: hello

UNCHANGED: world
"
`;
